### PR TITLE
[MLflow] Updating Pyfunc to specify resources needed to serve the model

### PR DESCRIPTION
**Pull request by sunishsheth2009**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/sunishsheth2009/mlflow/pull/11832?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11832/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11832
```

</p>
</details>

### What changes are proposed in this pull request?
[MLflow] Updating Pyfunc to specify resources needed to serve the model

TODO:
 - [ ] Update langchain to reuse resources class
 - [ ] Update langchain to not add resources in the flavor, rather add it as a top level entity

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Updating Pyfunc to specify resources needed to serve the model

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/pyfunc/__init__.py [hunk 1]
```diff
@@ -2202,6 +2205,11 @@ def predict(model_input: List[str]) -> List[str]:
             .. Note:: Experimental: This parameter may change or be removed in a future
                                     release without warning.
         example_no_conversion: {{ example_no_conversion }}
+        resources: A list of model resources or a resources.yaml file to list the resources
```
Comment (dbczumar) [inline mlflow/pyfunc/__init__.py:2208]:
```suggestion
        resources: A list of model resources or a resources.yaml file containing a list of resources
```

File: mlflow/models/model.py [hunk 2]
```diff
@@ -468,6 +470,23 @@ def model_size_bytes(self) -> Optional[int]:
...
+    @property
+    def resources(self) -> Optional[Dict[str, Any]]:
+        """
+        An optional dictionary that contains the resources required to serve the model.
+
+        :getter: Retrieves the resources required to serve the model
+        :setter: Sets the resources required to serve the model
+        :type: Optional[Dict[str, Any]]
+        """
+        return self._resources
```
Comment (dbczumar) [inline mlflow/models/model.py:483]: Can we make this an object - `ModelResources`?

**Review by dbczumar — APPROVED** [review]
LGTM - https://github.com/mlflow/mlflow/pull/11832/files#r1581641295. Can you also get approval from @harupy

File: mlflow/models/resources.py [hunk 3]
```diff
@@ -119,14 +118,16 @@ class _ResourceBuilder:
     def from_resources(
         resources: List[Resource], api_version: str = DEFAULT_API_VERSION
     ) -> Dict[str, Dict[ResourceType, List[Dict]]]:
-        resource_dict = defaultdict(lambda: defaultdict(list))
+        resource_dict = {}
```
Comment (BenWilson2) [inline mlflow/models/resources.py:121]: I prefer this :) Thanks

File: mlflow/java/scoring/src/main/java/org/mlflow/models/Model.java [hunk 4]
```diff
@@ -55,6 +55,12 @@ public static class Signature {
   @JsonProperty("metadata")
   private JsonNode metadata;
 
+  @JsonProperty("model_size_bytes")
```
Comment (BenWilson2) [inline mlflow/java/scoring/src/main/java/org/mlflow/models/Model.java:58]: Thanks for adding these

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks @sunishsheth2009

**Pull request merged**